### PR TITLE
LIME-266:  Connect OAuth to client

### LINKED ISSUE #266: Connect Connections Page to OAuth
Connect front-end Connections Page with OAuth functionality on the backend. When the user clicks on the "Connect/Disconnect" button send the appropriate request to the server

### PR DESCRIPTION
**Pull request by SofiiaTrokhymchuk**
Can't test the authorize/deauthorize button click flow because of CORS error
image
image
image
image

### CODE REVIEW

**Comment by anton-otroshchenko** [comment]
image
Try to initialize cors before other plugins because sometimes they may break it

**Comment by SofiiaTrokhymchuk** [comment]
Also doesn't work : (

image
image

**Comment by anton-otroshchenko** [comment]
Did you try to specify the origin url for strava?

**Comment by SofiiaTrokhymchuk** [comment]
If I specify origin and methods explicitly this also doesn't work
image

**Comment by anton-otroshchenko** [comment]
But you recieve your request not from our client but directly from strava or no?

**Comment by SofiiaTrokhymchuk** [comment]
image
image

**Comment by anton-otroshchenko** [comment]
fastify.register(cors, { origin: '*', credentials: true, }) 
try this

**Comment by anton-otroshchenko** [comment]
also check this please maybe it will help
https://groups.google.com/g/strava-api/c/YizXJN5EBM4
https://stackoverflow.com/questions/56461110/oauth2-httprequest-cors-policy

**Comment by SofiiaTrokhymchuk** [comment]
https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/93663923/899b2a12-3618-4b08-abf3-bfc3633c80c6

File: frontend/src/bundles/profile/pages/connections-page/connections-api.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    public constructor({ baseUrl, http, storage }: Constructor) {
+        super({ path: '', baseUrl, http, storage });
+
+        this.connectionsPath = ApiPath.CONNECTIONS;
+        this.oAuthPath = ApiPath.OAUTH;
+    }
+
+    public async getAll(): Promise<ConnectionGetAllItemResponseDto> {
+        const fullPath = `${this.connectionsPath}${ConnectionsPath.ROOT}`;
```
Comment (dimapopovych) [inline frontend/src/bundles/profile/pages/connections-page/connections-api.ts]: just pass parts to getFullEndpoint

File: frontend/src/bundles/profile/pages/connections-page/connections-api.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public async authorize(
+        provider: ValueOf<typeof OAuthProvider>,
+    ): Promise<void> {
+        const fullPath = `${this.oAuthPath}/${provider}${OAuthActionsPath.AUTHORIZE}`;
+
+        const response = await this.load(this.getFullEndpoint(fullPath, {}), {
+            method: 'GET',
+            contentType: ContentType.JSON,
+            hasAuth: true,
+        });
```
Comment (dimapopovych) [inline frontend/src/bundles/profile/pages/connections-page/connections-api.ts]: There is few such cases
```suggestion
        const response = await this.load(this.getFullEndpoint(this.oAuthPath, OAuthActionsPath.$PROVIDER_AUTORIZE, { provider }), {
            method: 'GET',
            contentType: ContentType.JSON,
            hasAuth: true,
        });
```

File: frontend/src/bundles/profile/pages/connections-page/components/connection-option.tsx [hunk 3]
```diff
@@ -1,27 +1,59 @@
...
+    }));
+
     const [isConnected, setIsConnected] = useState(false);
 
+    useEffect(() => {
+        const providerIsConnected = connections.find(
+            (connection) => connection.provider === provider,
+        );
+
+        setIsConnected(!!providerIsConnected);
```
Comment (dimapopovych) [inline frontend/src/bundles/profile/pages/connections-page/components/connection-option.tsx]:
```suggestion
        const existingConnection = connections.find(
            (connection) => connection.provider === provider,
        );

        setIsConnected(Boolean(existingConnection));
```

**Review by dimapopovych — CHANGES_REQUESTED** [review]
small changes required, probably could be merged after resolution
lgtm

**Comment by SofiiaTrokhymchuk** [comment]
I have also updated a little bit the `deauthorize` method to refresh the token if needed before the deauthorization, because I've found the bug that user can't deauthorize if token is expired and is not refreshed. Now everything works perfectly fine

**Review by nikita-remeslov — DISMISSED** [review]

**Review by dimapopovych — DISMISSED** [review]

**Review by dimapopovych — APPROVED** [review]

**Pull request merged**